monkey/monkey_island/cc/resources/agent_events.py

```diff
@@ -142,40 +148,50 @@ def _get_filtered_events(
         success: Optional[bool],
         timestamp_constraint: Optional[Tuple[str, float]],
     ) -> Sequence[AbstractAgentEvent]:
-        if type_ is not None:
-            events_by_type: Sequence[
-                AbstractAgentEvent
-            ] = self._agent_event_repository.get_events_by_type(type_)
-        else:
-            events_by_type = self._agent_event_repository.get_events()
-
-        if tag is not None:
-            events_by_tag: Sequence[
-                AbstractAgentEvent
-            ] = self._agent_event_repository.get_events_by_tag(tag)
+        if type_ is not None and tag is not None:
+            events = self._get_events_filtered_by_type_and_tag(type_, tag)
+        elif type_ is not None and tag is None:
+            events = self._agent_event_repository.get_events_by_type(type_)
+        elif type_ is None and tag is not None:
+            events = self._agent_event_repository.get_events_by_tag(tag)
         else:
-            events_by_tag = self._agent_event_repository.get_events()
-
-        # this has better time complexity than converting both lists to sets,
-        # finding their intersection, and then sorting the resultant set by timestamp
-        events = [event for event in events_by_tag if event in events_by_type]
+            events = self._agent_event_repository.get_events()
 
         if success is not None:
-            events = list(filter(lambda e: hasattr(e, "success") and e.success is success, events))  # type: ignore[attr-defined]  # noqa: E501
+            events = self._filter_events_by_success(events, success)
 
         if timestamp_constraint is not None:
-            operator, timestamp = timestamp_constraint
-            if operator == "gt":
-                separation_point = bisect_right(
-                    events, timestamp, key=lambda event: event.timestamp
-                )
-                events = events[separation_point:]
-            elif operator == "lt":
-                separation_point = bisect_left(events, timestamp, key=lambda event: event.timestamp)
-                events = events[:separation_point]
+            events = self._filter_events_by_timestamp(events, timestamp_constraint)
 
         return events
 
+    def _get_events_filtered_by_type_and_tag(
+        self, type_: Type[AbstractAgentEvent], tag: str
+    ) -> Sequence[AbstractAgentEvent]:
+        events_by_type = set(self._agent_event_repository.get_events_by_type(type_))
+        events_by_tag = set(self._agent_event_repository.get_events_by_tag(tag))
+
+        intersection = events_by_type.intersection(events_by_tag)
+        return sorted(intersection, key=lambda x: x.timestamp)
```

Comment:
It seems like this would be faster than doing the list comprehension. It was slow before because we executed it every time, and the both lists contained _all_ off the events (for GET agent-events). We aren't filtering by tag at this point, but I wonder how it will affect things when we do.